### PR TITLE
Fixes inconsistencies between `listen` and `emit`

### LINKED ISSUE #917: Semantic inconsistencies between `listen` and `emit` tasks
**What seems off**:

The `listen` task allows for user to configure events to listen to in the following way:

```yaml
listen:
  to:
    one:
      with: {}
```

Whereas the `emit` task does not define the `with` keyword:

```yaml
emit:
  event: {}
```

**What you expected to be**:

The `emit` task to define a `with` keyword, use to define event properties:

```yaml
emit:
  event:
    with: {}
```

This would allow for potential future properties such as `format`, and would make it more fluent, ubiquitous and consistent.

### PR DESCRIPTION
**Pull request by JBBianchi**
**Please specify parts of this PR update:**

- [x] Specification
- [x] Schema
- [ ] Examples
- [ ] Extensions
- [ ] Use Cases
- [ ] Community
- [x] CTK
- [ ] Other

**Discussion or Issue link**:
Closes #917 (and other little minor typo)

**What this PR does**:
- Fixes inconsistencies between `listen` and `emit`
  - Added new `event` section in the specification documentation
  - Fixed the sample of `emit` in the specification documentation
  - Added new `event` definition in the JSON schema and referenced it in `emit.event.with` and `eventFilter.with`
- Fixes a typo in the set feature
- Fixed type of an http call `endpoint` by adding `string`
- Added new (missing) `endpoint` section in the specification documentation
- Escaped * character
- Fixed double `https://https://`
- Fixed links to runtime expressions
- Fixed typo in `implictly`

**Additional information:**
\-

### CODE REVIEW

File: dsl-reference.md [hunk 1]
```diff
@@ -1389,6 +1391,25 @@ Represents the configuration of an event consumption strategy.
...
+#### Properties
+
+| Property | Type | Required | Description |
+|----------|:----:|:--------:|-------------|
+| id | `string` | `no` | Identifies the event. `source` + `id` is unique for each distinct event.<br>*Required when emitting an event using `emit.event.with`.* |
+| source | `string` | `no` | An URI formatted string that identifies the context in which an event happened. `source` + `id` is unique for each distinct event.<br>*Required when emitting an event using `emit.event.with`.* |
+| type | `string` | `no` | Describes the type of event related to the originating occurrence. `source` + `id` is unique for each distinct event.<br>*Required when emitting an event using `emit.event.with`.* |
+| time | `string` | `no` | The timestamp of when the occurrence happened. |
+| subject | `string` | `no` | Describes the subject of the event in the context of the event producer. |
+| datacontenttype | `string` | `no` | Content type of `data` value. If omitted, it implies the `data` is a JSON value confiorming to the "application/json" media type. |
```
Comment (cdavernas) [inline dsl-reference.md:1407]: Typo here: confiorming instead of conforming

**Review by cdavernas — CHANGES_REQUESTED** [review]
Looks great to me, thank you so much!

Aside from a small typo, I'm wondering whether the term `event` applies to `with`: IMHO, we should find a term that best describes it. It's as a matter of fact not an `event`, which is a term that could confuse users, but rather `eventProperties` or something like that. WDYT?

**Review by cdavernas — APPROVED** [review]
Thank you so much bro! ❤️

**Comment by ricardozanini** [comment]
@JBBianchi can you please rebase?

File: dsl-reference.md [hunk 2]
```diff
@@ -368,7 +369,7 @@ The [HTTP Call](#http-call) enables workflows to interact with external services
 | Name | Type | Required | Description|
 |:--|:---:|:---:|:---|
 | method | `string` | `yes` | The HTTP request method. |
-| endpoint | [`endpoint`](#endpoint) | `yes` | An URI or an object that describes the HTTP endpoint to call. |
+| endpoint | `string`\|[`endpoint`](#endpoint) | `yes` | An URI or an object that describes the HTTP endpoint to call. |
```
Comment (ricardozanini) [inline dsl-reference.md:372]: Why do we need this `\` scape char here?

File: dsl-reference.md [hunk 3]
```diff
@@ -528,9 +530,9 @@ Allows workflows to iterate over a collection of items, executing a defined set
 | Name | Type | Required | Description|
 |:--|:---:|:---:|:---|
 | for.each | `string` | `no` | The name of the variable used to store the current item being enumerated.<br>Defaults to `item`. |
-| for.in | `string` | `yes` | A [runtime expression](#runtime-expressions) used to get the collection to enumerate. |
+| for.in | `string` | `yes` | A [runtime expression](dsl.md##runtime-expressions) used to get the collection to enumerate. |
```
Comment (ricardozanini) [inline dsl-reference.md:533]: Do we need the double anchor?

File: dsl-reference.md [hunk 4]
```diff
@@ -528,9 +530,9 @@ Allows workflows to iterate over a collection of items, executing a defined set
...
 | for.at | `string` | `no` | The name of the variable used to store the index of the current item being enumerated.<br>Defaults to `index`. |
-| while | `string` | `no` | A [runtime expression](#runtime-expressions) that represents the condition, if any, that must be met for the iteration to continue. |
+| while | `string` | `no` | A [runtime expression](dsl.md##runtime-expressions) that represents the condition, if any, that must be met for the iteration to continue. |
```
Comment (ricardozanini) [inline dsl-reference.md:535]: Same here

File: dsl-reference.md [hunk 5]
```diff
@@ -1389,6 +1391,28 @@ Represents the configuration of an event consumption strategy.
 | any | [`eventFilter[]`](#event-filter) | `no` | Configures the workflow to wait for any of the defined events before resuming execution.<br>*Required if `all` and `one` have not been set.* |
 | one | [`eventFilter`](#event-filter) | `no` | Configures the workflow to wait for the defined event before resuming execution.<br>*Required if `all` and `any` have not been set.* |
 
+### Event Properties
+
+An event object typically includes details such as the event type, source, timestamp, and unique identifier, along with any relevant data payload. The [Cloud Events specification](https://cloudevents.io/), favored by Serveless Workflow, standardizes this structure to ensure interoperability across different systems and services.
```
Comment (ricardozanini) [inline dsl-reference.md:1396]:
```suggestion
An event object typically includes details such as the event type, source, timestamp, and unique identifier along with any relevant data payload. The [Cloud Events specification](https://cloudevents.io/), favored by Serverless Workflow, standardizes this structure to ensure interoperability across different systems and services.
```

File: dsl-reference.md [hunk 6]
```diff
@@ -1389,6 +1391,28 @@ Represents the configuration of an event consumption strategy.
...
+
+#### Properties
+
+| Property | Type | Required | Description |
+|----------|:----:|:--------:|-------------|
+| id | `string` | `no` | Identifies the event. `source` + `id` is unique for each distinct event.<br>*Required when emitting an event using `emit.event.with`.* |
+| source | `string` | `no` | An URI formatted string that identifies the context in which an event happened. `source` + `id` is unique for each distinct event.<br>*Required when emitting an event using `emit.event.with`.* |
+| type | `string` | `no` | Describes the type of event related to the originating occurrence. `source` + `id` is unique for each distinct event.<br>*Required when emitting an event using `emit.event.with`.* |
```
Comment (ricardozanini) [inline dsl-reference.md:1404]:
```suggestion
| type | `string` | `no` | Describes the type of event related to the originating occurrence. <br>*Required when emitting an event using `emit.event.with`.* |
```

File: dsl-reference.md [hunk 7]
```diff
@@ -1389,6 +1391,28 @@ Represents the configuration of an event consumption strategy.
...
+| time | `string` | `no` | The timestamp of when the occurrence happened. |
+| subject | `string` | `no` | Describes the subject of the event in the context of the event producer. |
+| datacontenttype | `string` | `no` | Content type of `data` value. If omitted, it implies the `data` is a JSON value conforming to the "application/json" media type. |
+| dataschema | `string` | `no` | An URI that identifies the schema that `data` adheres to. |
+| data | `object` | `no` | The event payload. |
+
+*Additional properties can be supplied.*
```
Comment (ricardozanini) [inline dsl-reference.md:1411]: Worth linking to https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#extension-context-attributes. Any properties not listed in the CE spec are considered to be extensions.

File: dsl-reference.md [hunk 8]
```diff
@@ -1473,7 +1497,7 @@ When set, runtimes must validate input data against the defined schema, unless d
 | Property | Type | Required | Description |
 |----------|:----:|:--------:|-------------|
 | schema | [`schema`](#schema) | `no` | The [`schema`](#schema) used to describe and validate input data.<br>*Even though the schema is not required, it is strongly encouraged to document it, whenever feasible.* |
-| from | `string`<br>`object` | `no` | A [runtime expression](#runtime-expressions), if any, used to filter and/or mutate the workflow/task input.  |
+| from | `string`<br>`object` | `no` | A [runtime expression](dsl.md##runtime-expressions), if any, used to filter and/or mutate the workflow/task input.  |
```
Comment (ricardozanini) [inline dsl-reference.md:1500]:
```suggestion
| from | `string`<br>`object` | `no` | A [runtime expression](dsl.md#runtime-expressions), if any, used to filter and/or mutate the workflow/task input.  |
```

File: dsl-reference.md [hunk 9]
```diff
@@ -1397,7 +1421,7 @@ An event filter is a mechanism used to selectively process or handle events base
 
 | Property | Type | Required | Description |
 |----------|:----:|:--------:|-------------|
-| with | `object` | `yes` | A name/value mapping of the attributes filtered events must define. Supports both regular expressions and runtime expressions.  |
+| with | [`eventProperties`](#event-properties) | `yes` | A name/value mapping of the attributes filtered events must define. Supports both regular expressions and runtime expressions.  |
```
Comment (ricardozanini) [inline dsl-reference.md:1424]: So this type in the schema can be the new definition introduced by @matthias-pichler-warrify and also a [pattern](https://json-schema.org/understanding-json-schema/reference/string#regexp)? Maybe it's worth updating the schema?

File: dsl-reference.md [hunk 10]
```diff
@@ -1502,7 +1526,7 @@ When set, runtimes must validate output data against the defined schema, unless
 | Property | Type | Required | Description |
 |----------|:----:|:--------:|-------------|
 | schema | [`schema`](#schema) | `no` | The [`schema`](#schema) used to describe and validate output data.<br>*Even though the schema is not required, it is strongly encouraged to document it, whenever feasible.* |
-| as | `string`<br>`object` | `no` | A [runtime expression](#runtime-expressions), if any, used to filter and/or mutate the workflow/task output. |
+| as | `string`<br>`object` | `no` | A [runtime expression](dsl.md##runtime-expressions), if any, used to filter and/or mutate the workflow/task output. |
```
Comment (ricardozanini) [inline dsl-reference.md:1529]:
```suggestion
| as | `string`<br>`object` | `no` | A [runtime expression](dsl.md#runtime-expressions), if any, used to filter and/or mutate the workflow/task output. |
```

File: schema/workflow.yaml [hunk 11]
```diff
@@ -830,28 +837,8 @@ $defs:
     properties:
       with:
         title: WithEvent
-        type: object
+        $ref: '#/$defs/eventProperties'
```
Comment (ricardozanini) [inline schema/workflow.yaml:840]: This one could be a `OneOf` of expression, pattern, or `eventProperties`.

**Review by ricardozanini — COMMENTED** [review]
Great work, @JBBianchi! Just left a few comments if you don't mind. Also, do you mind adding a few examples to validate the change? Thanks!

File: dsl-reference.md (hunk 2 above)
Comment (JBBianchi) [reply dsl-reference.md:372]: Otherwise it breaks the table formatting (like adding an extra cell).

File: dsl-reference.md (hunk 3 above)
Comment (JBBianchi) [reply dsl-reference.md:533]: It's a mistake, good catch.

File: schema/workflow.yaml (hunk 11 above)
Comment (JBBianchi) [reply schema/workflow.yaml:840]: I'm not sure about that one. @cdavernas @fjtirado @matthias-pichler-warrify wdyt ?

**Pull request closed without merging**